Generic training script

GMW99 opened the pull request.

Commit:
Initial training file.

Commit:
Move the callbacks classes

Separating out the notebook such that a training script can be easily written a modified in the future.

Commit:
Moving the kompressor classes

Separating out the notebook such that a training script can be easily written a modified in the future.

Commit:
Moving the model functions

Separating out the notebook such that a training script can be easily written a modified in the future.

Commit:
Moving the predictor functions

Separating out the notebook such that a training script can be easily written a modified in the future.

Commit:
Creating a training script

This is a script that will allow a dataset to be passed along with the relevant parameters.

Commit:
Creating Plugin Method for Models

So that different Models can be selected and used for training a plugin method is being implemented.

Commit:
Creating Plugin Method for Models

So that different Models can be selected and used for training a plugin method is being implemented.

Commit:
Adding named_module function

Commit:
Plugin method for predictors

This allows for multiple predictors to be selected.

JossWhittle commented on training/kompressors/haiku.py:
```diff
@@ -0,0 +1,80 @@
+import jax.numpy as jnp
+from tqdm import trange
+
+class HaikuKompressor(Kompressor):
```
Can you add save and load functions for the model weights? 

JAX jnp.array objects need to be cast into np.array to be saved and then cast back to jnp.array after being loaded.

This will be useful for training a baseline model so we can encode the maps for a dataset at a scale, then start benchmarking secondary compressors side by side.

JossWhittle commented on training/models/__init__.py:
```diff
@@ -0,0 +1,29 @@
+AVAILABLE = ['cnn', 'mlp']
+Model = namedtuple('Model', ['name', "Model"])
+
+def use_model(model):
```
I like this pattern, but is there a library that handles/abstracts this pattern already? Something where we put a decorator on the class definition to say it will be an model option? 

Happy to go this route for now, especially to get something prototyped and working.

GMW99 replied:
After a quick search, I have this library that looks actively maintained https://github.com/pytest-dev/pluggy/ (last updated 4 days ago and is used in pytest and tox and devpi), its worth a look as it may make development easier in the future.

JossWhittle commented on training/train.py:
```diff
@@ -0,0 +1,15 @@
+    parser = argparse.ArgumentParser(description="Setup training")
+    parser.add_argument(
+        "-l", "--level", type=int, default=1, help=""
+    )
```
rest-of-the-owl.jpg ;)

JossWhittle commented on training/callbacks/base.py:
```diff
@@ -0,0 +1,12 @@
+from abc import abstractmethod, ABCMeta
```
I always forget to use this! Thank you

Commit:
Renamed to make clearer.

JossWhittle reviewed (commented):
Looking good. I like the registered classes pattern, I just wonder if there is an existing framework to apply that pattern instead of having the utility functions for it as part of the compressor code.

Commit:
Plugin method for compressors

This allows for different compressors to be selected/used.

Commit:
Adding arguments

Commit:
Parsing arguments and setting up plugins

Commit:
Added chunking and logging arguments

Commit:
Adding training section

This is taken from the notebook, but modified to support the arguments passed from the user.

Commit:
Sorting out defaults values for arguments.

Commit:
Sorting out module calling route to the abstract base class

Commit:
Adding abstractmethods

Methods for fit,save and load model.

Commit:
Implementation of save_model and load_model

Commit:
Options for saving and loading models added

Commit:
formatting

Commit:
Changed Metrics to match changes made in SHA: 5184e49d68f2c336d3a2383c1632bbc0d9d56c8f

Modified so it iterates over the maps for where the keys intersect.

Commit:
Moved training files to src/kompressor

This has been done so the modules are built within the src.

Commit:
Moved training files to src/kompressor

This has been done so the modules are built within the src.

Commit:
Moved training files to src/kompressor

This has been done so the modules are built within the src.
